lib/mongodb/native_mongodb.nit

```diff
@@ -0,0 +1,492 @@
+	fun destroy `{ bson_destroy(recv); `}
+
+	# Utility method to set `Sys.last_mongoc_error`.
+	fun set_mongoc_error(err: BSONError) do sys.last_mongoc_error = err
+end
+
+# Wrapper for `bson_error_t`.
+#
+# The `bson_error_t` structure is used to encapsulate information about an error.
+#
+# See [`bson_error_t`](http://api.mongodb.org/libbson/current/bson_error_t.html).
+extern class BSONError `{ bson_error_t * `}
+
+	# Wrapper for `error.domain`.
+	#
+	# The `error.domain` field contains the logical domain within a library that
+	# created the error.
+	fun domain: Int `{ return recv->domain; `}
+
+	# Wrapper for `error.code`.
+	#
+	# The `error.code` field contains the domain specific error code.
+	fun code: Int `{ return recv->code; `}
+
+	# Wrapper for `error.message`.
+	#
+	# The `error.message` field contains a human printable error message.
+	fun message: NativeString `{ return recv->message; `}
+end
+
+redef class Sys
+	# Last error raised by `monogdb::MongoClient`.
+	#
+	# See `MongoClient::last_error`.
+	var last_mongoc_error: nullable BSONError = null
+end
+
+# Wrapper for `char**`.
+#
+# Used to handle array of NativeString returned by MongoDB.
+redef class NativeCStringArray
+	# Frees `self`.
+	#
+	# This instance should not be used beyond this point!
+	fun destroy `{ free(recv); `}
+end
+
+# Wrapper for `mongoc_client_t`.
+#
+# `mongoc_client_t` is an opaque type that provides access to a MongoDB node,
+# replica-set, or sharded-cluster.
+# It maintains management of underlying sockets and routing to individual nodes.
+#
+# See [`mongoc_client_t`](http://api.mongodb.org/c/current/mongoc_client_t.html).
+extern class NativeMongoClient `{ mongoc_client_t * `}
+
+	# Wrapper for `mongoc_client_new()`.
+	#
+	# Creates a new `mongoc_client_t` using the `uri` string provided.
+	new(uri: NativeString) `{
+		mongoc_init();
+		return mongoc_client_new(uri);
+	`}
+
+	# Wrapper for `mongoc_client_get_server_status()`.
+	#
+	# Queries the server for the current server status.
+	# Returns `null` if an error occured.
+	fun server_status: nullable NativeBSON import set_mongoc_error, NativeBSON.as nullable `{
+		bson_error_t error;
+		bson_t *reply = bson_new();
+		if(!mongoc_client_get_server_status(recv, NULL, reply, &error)){
+			NativeMongoClient_set_mongoc_error(recv, &error);
+			return null_NativeBSON();
+		}
+		return NativeBSON_as_nullable(reply);
+	`}
+
+	# Wrapper for `mongoc_client_get_database_names()`.
+	#
+	# This function queries the MongoDB server for a list of known databases.
+	# Returns `null` if an error occured.
+	fun database_names: nullable NativeCStringArray
+		import set_mongoc_error, NativeCStringArray, NativeCStringArray.as nullable `{
+		bson_error_t error;
+		char **strv;
+		if(strv = mongoc_client_get_database_names(recv, &error)) {
+			return NativeCStringArray_as_nullable(strv);
+		}
+		NativeMongoClient_set_mongoc_error(recv, &error);
+		return null_NativeCStringArray();
```

Comment:
It's OK as is, but for the next time when dealing with extern instances you can use the C NULL. Then, on the Nit side you can use `address_is_null`. It's much cleaner and as it doesn't use cast callbacks, it only needs the light FFI.